src/Function.cpp

```diff
@@ -248,6 +338,109 @@ Function::Function(const std::string &n) : contents(new FunctionContents) {
+    }
+
+    IntrusivePtr<FunctionContents> &copied_func = copied_map[src.func];
+    if (copied_func.defined()) {
+        dst.func = copied_func;
+    } else {
+        dst.func = IntrusivePtr<FunctionContents>(new FunctionContents);
+        deep_copy_function_contents_helper(dst.func, src.func, copied_map);
+        copied_map[src.func] = dst.func;
+    }
+}
+
+IntrusivePtr<FunctionContents> deep_copy_function_contents_helper(
+        const IntrusivePtr<FunctionContents> &src, DeepCopyMap &copied) {
+
+    IntrusivePtr<FunctionContents> copy(new FunctionContents);
+    deep_copy_function_contents_helper(copy, src, copied);
+    return copy;
+}
+
+void deep_copy_function_contents_helper(IntrusivePtr<FunctionContents> &dst,
+                                        const IntrusivePtr<FunctionContents> &src,
+                                        DeepCopyMap &copied_map) {
+    debug(4) << "Deep-copy function contents: \"" << src.ptr->name << "\"\n";
+
+    internal_assert(dst.defined() && src.defined()) << "Cannot deep-copy undefined Function\n";
+
+    dst.ptr->name = src.ptr->name;
+    dst.ptr->args = src.ptr->args;
+    dst.ptr->values = src.ptr->values;
+    dst.ptr->output_types = src.ptr->output_types;
+    dst.ptr->debug_file = src.ptr->debug_file;
+    dst.ptr->extern_function_name = src.ptr->extern_function_name;
+    dst.ptr->extern_is_c_plus_plus = src.ptr->extern_is_c_plus_plus;
+    dst.ptr->trace_loads = src.ptr->trace_loads;
+    dst.ptr->trace_stores = src.ptr->trace_stores;
+    dst.ptr->trace_realizations = src.ptr->trace_realizations;
+    dst.ptr->frozen = src.ptr->frozen;
+    dst.ptr->output_buffers = src.ptr->output_buffers;
+
+    dst.ptr->schedule = src.ptr->schedule.deep_copy(copied_map);
+
+    for (const auto &u : src.ptr->updates) {
+        UpdateDefinition u_copy;
+        deep_copy_update_definition_helper(dst, u_copy, u, copied_map);
+        internal_assert(u_copy.domain.same_as(u_copy.schedule.reduction_domain()))
+            << "UpdateDefinition should point to the same reduction domain as its schedule\n";
+        dst.ptr->updates.push_back(std::move(u_copy));
+    }
+    for (const auto &e : src.ptr->extern_arguments) {
+        ExternFuncArgument e_copy;
+        deep_copy_extern_func_argument_helper(e_copy, e, copied_map);
+        dst.ptr->extern_arguments.push_back(std::move(e_copy));
+    }
+}
+
+void Function::deep_copy(Function &copy, std::map<Function, Function> &copied_map) const {
+    internal_assert(copy.contents.defined() && contents.defined()) << "Cannot deep-copy undefined Function\n";
+    DeepCopyMap copied_funcs_map;
+    for (const auto &iter : copied_map) {
+        copied_funcs_map[iter.first.contents] = iter.second.contents;
+    }
+    // Add reference to its copy in case of self-reference.
+    copied_funcs_map[contents] = copy.contents;
+
+    deep_copy_function_contents_helper(copy.contents, contents, copied_funcs_map);
+
+    for (const auto &iter : copied_funcs_map) {
```

Comment:
Had to stare this code for a bit to understand it. This function could use a few more comments explaining what it's doing.